File: src/windows/AmbientContextMcp.Core.Tests/PolicyVersionTests.cs
using AmbientContextMcp.Core.Hub;
using AmbientContextMcp.Core.Models;
using Xunit;

namespace AmbientContextMcp.Core.Tests;

public class PolicyVersionTests
{
    [Fact]
    public void Same_policy_produces_same_version()
    {
        var classifications = new List<PrivacyClassification>
        {
            new() { Path = "media.title", Sensitivity = "high", DefaultTransmit = false }
        };
        var overrides = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase)
        {
            { "media.title", true }
        };

        var a = PolicyVersionService.ComputePolicyVersion(classifications, overrides);
        var b = PolicyVersionService.ComputePolicyVersion(classifications, overrides);

        Assert.False(string.IsNullOrEmpty(a));
        Assert.Equal(a, b);
    }

    [Fact]
    public void Changing_override_changes_version()
    {
        var classifications = new List<PrivacyClassification>
        {
            new() { Path = "media.title", Sensitivity = "high", DefaultTransmit = false }
        };

        var beforeOverrides = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase);
        var afterOverrides = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase)
        {
            { "media.title", true }
        };

        var before = PolicyVersionService.ComputePolicyVersion(classifications, beforeOverrides);
        var after = PolicyVersionService.ComputePolicyVersion(classifications, afterOverrides);

        Assert.NotEqual(before, after);
    }

    [Fact]
    public void Override_order_does_not_affect_version()
    {
        var classifications = new List<PrivacyClassification>
        {
            new() { Path = "media.title", Sensitivity = "high", DefaultTransmit = false }
        };
        var ordered1 = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase)
        {
            { "media.title", true },
            { "media.artist", true }
        };
        var ordered2 = new Dictionary<string, bool>(StringComparer.OrdinalIgnoreCase)
        {
            { "media.artist", true },
            { "media.title", true }
        };

        Assert.Equal(
            PolicyVersionService.ComputePolicyVersion(classifications, ordered1),
            PolicyVersionService.ComputePolicyVersion(classifications, ordered2));
    }

    [Fact]
    public void Hub_exposes_policy_version_via_state_and_poll_responses()
    {
        var hub = LocalContextHubTestFactory.CreateInMemory();
        hub.Ingest(new AmbientContextSnapshot
        {
            ObservedAt = DateTimeOffset.UtcNow,
            PrivacyClassifications =
            [
                new() { Path = "media.title", Sensitivity = "high", DefaultTransmit = false }
            ]
        });

        var state = hub.GetContextStates(new LocalContextStateRequest());
        var poll = hub.PollEvents(new LocalContextPollRequest { ClientId = "test" });

        Assert.False(string.IsNullOrEmpty(state.PolicyVersion));
        Assert.Equal(state.PolicyVersion, poll.PolicyVersion);
    }
}

File: src/macos/Tests/AmbientContextCoreTests/PolicyVersionTests.swift
import Foundation
import Testing
@testable import AmbientContextCore

@Suite("PolicyVersion")
struct PolicyVersionTests {
    private static let classifications = [
        PrivacyClassification(path: "media.title", sensitivity: "high", defaultTransmit: false)
    ]

    @Test("Same_policy_produces_same_version")
    func samePolicyProducesSameVersion() {
        let overrides: CaseInsensitiveDictionary<Bool> = ["media.title": true]

        let a = PolicyVersionService.computePolicyVersion(
            classifications: Self.classifications, overrides: overrides)
        let b = PolicyVersionService.computePolicyVersion(
            classifications: Self.classifications, overrides: overrides)

        #expect(!a.isEmpty)
        #expect(a == b)
    }

    @Test("Changing_override_changes_version")
    func changingOverrideChangesVersion() {
        let before = PolicyVersionService.computePolicyVersion(
            classifications: Self.classifications, overrides: [:])
        let after = PolicyVersionService.computePolicyVersion(
            classifications: Self.classifications, overrides: ["media.title": true])

        #expect(before != after)
    }

    @Test("Override_order_does_not_affect_version")
    func overrideOrderDoesNotAffectVersion() {
        let ordered1: CaseInsensitiveDictionary<Bool> = ["media.title": true, "media.artist": true]
        let ordered2: CaseInsensitiveDictionary<Bool> = ["media.artist": true, "media.title": true]

        #expect(
            PolicyVersionService.computePolicyVersion(
                classifications: Self.classifications, overrides: ordered1) ==
            PolicyVersionService.computePolicyVersion(
                classifications: Self.classifications, overrides: ordered2))
    }

    @Test("Hub_exposes_policy_version_via_state_and_poll_responses")
    func hubExposesPolicyVersionViaStateAndPollResponses() {
        let hub = LocalContextHubTestFactory.createInMemory()
        hub.ingest(AmbientContextSnapshot(
            observedAt: Date(),
            privacyClassifications: [
                PrivacyClassification(path: "media.title", sensitivity: "high", defaultTransmit: false)
            ]))

        let state = hub.getContextStates(LocalContextStateRequest())
        let poll = hub.pollEvents(LocalContextPollRequest(clientId: "test"))

        #expect(!state.policyVersion.isEmpty)
        #expect(state.policyVersion == poll.policyVersion)
    }
}
